base/inference.jl

```diff
@@ -1479,7 +1529,20 @@ function typeinf(linfo::LambdaStaticData,atypes::Tuple,sparams::Tuple, def, cop)
                 end
             end
             pc´ = pc+1
-            if isa(stmt,GotoNode)
+            if isa(changes,StateUpdate) && isa((changes::StateUpdate).var, GenSym)
+                changes = changes::StateUpdate
+                id = (changes.var::GenSym).id+1
+                new = changes.vtype
+                old = gensym_types[id]
+                if old===NF || !(new <: old)
+                    gensym_types[id] = tmerge(old, new)
+                    for r in gensym_uses[id]
+                        if !is(s[r],()) # s[r] === () => unreached statement
```

Comment:
`s[r] === ()` doesn't mean `r` is never reached; it means it hasn't been reached *yet*.

`W` keeps track of control flow. You should not visit all places a variable is used when its type changes. You should only visit those places that the information actually reaches via control flow.

A further problem is that we will want to generalize this algorithm to derive type information from branches: inside `if isa(x,T)` we want to know `x::T`. So we should continue to track types for all variables at all program points.